### PR TITLE
Improper basic steps implementation

### PR DESCRIPTION
**Pull request by qtothec**
## Summary/Motivation:
This PR introduces the ability to handle improper basic steps to the basic steps implementation in Pyomo.GDP. Tests are also added to exercise this ability.

Depends on #581 because of testing example.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/gdp/basic_step.py [hunk 1]
```diff
@@ -26,10 +30,23 @@ def _pseudo_clone(self):
...
 
 
-def apply_basic_step( disjunctions ):
+def _squish_singletons(tuple_iter):
+    """Squish all singleton tuples into their non-tuple values."""
+    for tup in tuple_iter:
+        if len(tup) == 1:
+            yield tup[0]
+        else:
+            yield tup
```
Comment (jsiirola) [inline pyomo/gdp/basic_step.py]: Why not use `flatten_tuple`?

File: pyomo/gdp/basic_step.py [hunk 2]
```diff
@@ -57,12 +74,20 @@ def apply_basic_step( disjunctions ):
...
                     continue
                 tmp.del_component(k)
                 ans.disjuncts[idx].src[i].add_component(k,v)
+        # Copy in the constraints corresponding to the improper disjunctions
+        ans.disjuncts[idx].improper_constraints = ConstraintList()
+        for constr in constraints:
+            for indx in constr:
+                ans.disjuncts[idx].improper_constraints.add(
+                    (constr[indx].lower, constr[indx].body, constr[indx].upper)
+                )
```
Comment (jsiirola) [inline pyomo/gdp/basic_step.py]: Does it make sense to give the user an option to deactivate the original constraint after applying the basic step?

Comment (qtothec) [reply pyomo/gdp/basic_step.py]: Do I not deactivate the original constraint? If so, that's an oversight on my part.

File: pyomo/gdp/basic_step.py (hunk 1 above)
Comment (qtothec) [reply pyomo/gdp/basic_step.py]: Actually, looking at what `flatten_tuple` does, we only want to flatten it if it's a singleton. That is, `((0, 0), (0, 1), (0, 2))` is fine, but when only one disjunction is participating, it becomes `((0,), (1,), (2,))`, which is problematic as input to `initialize=` for `Set(dimen=1)`.

**Pull request merged**